intellij-toml/schema/src/main/kotlin/org/toml/ide/schema/TomlJsonSchemaCompletionContributor.kt

```diff
@@ -0,0 +1,141 @@
+
+                    val schemaProperties = schema.properties
+                    addAllPropertyVariants(properties, adapter, schemaProperties, knownNames)
+                }
+            }
+
+            for (variant in variants) {
+                resultConsumer.consume(variant)
+            }
+        }
+
+        private fun addAllPropertyVariants(
+            properties: Collection<String>,
+            adapter: JsonPropertyAdapter?,
+            schemaProperties: Map<String, JsonSchemaObject>,
+            knownNames: MutableSet<String>
+        ) {
+            val variants = schemaProperties.keys.filter { name ->
+                !properties.contains(name) && !knownNames.contains(name) || name == adapter?.name
+            }
+
+            for (variant in variants) {
+                knownNames.add(variant)
+                addPropertyVariant(variant, schemaProperties[variant])
+            }
+        }
+
+        @Suppress("NAME_SHADOWING")
+        private fun addPropertyVariant(key: String, jsonSchemaObject: JsonSchemaObject?) {
+            val currentVariants = JsonSchemaResolver(project, jsonSchemaObject!!).resolve()
+            val jsonSchemaObject = ObjectUtils.coalesce(ContainerUtil.getFirstItem(currentVariants), jsonSchemaObject)
+
+            var description = JsonSchemaDocumentationProvider.getBestDocumentation(true, jsonSchemaObject)
+            if (description.isNullOrBlank()) {
+                description = jsonSchemaObject.getTypeDescription(true).orEmpty()
+            }
+
+            val lookupElement = LookupElementBuilder.create(key)
+                .withTypeText(description)
+                .withIcon(getIconForType(jsonSchemaObject.guessType()))
+
+            variants.add(lookupElement)
+        }
+
+        private fun getIconForType(type: JsonSchemaType?) = when (type) {
+            null -> AllIcons.Nodes.Property
+            JsonSchemaType._object -> AllIcons.Json.Object
+            JsonSchemaType._array -> AllIcons.Json.Array
+            else -> AllIcons.Nodes.Property
+        }
+    }
+
+    companion object {
+        fun doCompletion(parameters: CompletionParameters, result: CompletionResultSet, rootSchema: JsonSchemaObject, stop: Boolean) {
+            val completionPosition = parameters.originalPosition ?: parameters.position
+            val worker = Worker(rootSchema, parameters.position, completionPosition, result)
+
+            worker.work()
+
+            if (stop && worker.variants.isNotEmpty()) {
+                result.stopHere()
+            }
+        }
+
+        @TestOnly
+        fun getCompletionVariants(
+            schema: JsonSchemaObject,
+            position: PsiElement,
+            originalPosition: PsiElement
+        ): List<LookupElement> {
+            val result = mutableListOf<LookupElement>()
+            Worker(schema, position, originalPosition) { element: LookupElement? -> result.add(element!!) }.work()
```

Comment:
I understand that we will drop this code but anyway: let's try to avoid `!!` in code
If we really expect not null object somewhere but can't express it via type system, use `error` with an explicit message